lib/cereal/cereal.go

```diff
@@ -497,19 +503,188 @@ type TaskExecutorOpts struct {
+// - in-process notifications that are sent when we think there
+// might be a task available.
+//
+// When either of these tiggers occur, we spawn a new goroutine (the
+// "task worker") provided that we are under the maximum configured
+// workers for this task executor.
+//
+// The task worker will repeatedly dequeue tasks until none are
+// available. To ensure our queue grows in response to load, if a task
+// is found in the queue, another task worker will be spawned.
+//
 type registeredExecutor struct {
+	name     string
 	executor TaskExecutor
 	opts     TaskExecutorOpts
+
+	wakeupChan chan struct{}
+
+	maxWorkers    int
+	activeWorkers int
+	sync.Mutex    // guards activeWorkers
+
+	wg sync.WaitGroup
+}
+
+// DecActiveWorkers decrements the worker count, clamping it at zero.
+func (r *registeredExecutor) DecActiveWorkers() {
+	r.Lock()
+	defer r.Unlock()
+
+	if r.activeWorkers > 0 {
+		r.activeWorkers--
+	}
+}
+
+// IncActiveWorkers increments the active worker count if another
+// worker is allowed. It returns true if the count was incremented and
+// false otherwise. The caller should not start a worker if false is
+// returned.
+func (r *registeredExecutor) IncActiveWorkers() bool {
+	r.Lock()
+	defer r.Unlock()
+
+	if r.activeWorkers < r.maxWorkers {
+		r.activeWorkers++
+		return true
+	}
+
+	return false
+}
+
+// WakeupPoller will wake up the poller for this
+// registeredExecutor. This is called when a workflow executor
+// enqueues tasks.
+func (r *registeredExecutor) WakeupPoller() {
+	select {
+	case r.wakeupChan <- struct{}{}:
+	default:
+	}
+}
+
+// StartPoller runs until the given context is cancelled, starting a
+// TaskWorker every interval (or sooner if we've been notified of a
+// change).
+func (r *registeredExecutor) StartPoller(ctx context.Context, b backend.Driver, taskPollInterval time.Duration, workflowWakeupFun func()) {
+	logctx := logrus.WithField("task_name", r.name)
+	logctx.Infof("Starting task poller")
+	for {
+		select {
+		case <-ctx.Done():
+			logctx.WithField("active_workers", r.activeWorkers).Info("Waiting for all task workers to exit")
+			r.wg.Wait()
+			logctx.Info("Exiting task poller")
+			return
+		case <-time.After(taskPollInterval):
+			r.startTaskWorker(ctx, b, workflowWakeupFun)
+		case <-r.wakeupChan:
+			r.startTaskWorker(ctx, b, workflowWakeupFun)
+		}
+	}
+}
+
+// startTaskWorker starts a goroutine that will dequeue new tasks and
+// execute them, exiting when no tasks are available in the queue. If
+// the task poller initially finds a job, it will also start the next
+// task worker.
+func (r *registeredExecutor) startTaskWorker(ctx context.Context, b backend.Driver, workflowWakeupFun func()) {
+	logctx := logrus.WithField("task_name", r.name)
+	if !r.IncActiveWorkers() {
+		logctx.WithFields(logrus.Fields{
+			"max_workers":    r.maxWorkers,
+			"active_workers": r.activeWorkers,
+		}).Warn("Maximum task workers already started")
+		return
+	}
+
+	go func() {
+		r.wg.Add(1)
```

Comment:
i think this needs to happen before creating to goroutine

Reply:
Good catch. this could race with a concurrently running Wait()